sdk/trace/src/main/scala/org/typelevel/otel4s/sdk/trace/SdkSpanBackend.scala

```diff
@@ -0,0 +1,311 @@
+  *     and so on
+  *   - mutable state - can be modified during the lifecycle, for example: name,
+  *     attributes, events, etc
+  *
+  * All modifications of the mutable state are ignored once the span is ended.
+  *
+  * @see
+  *   [[https://opentelemetry.io/docs/specs/otel/trace/api]]
+  *
+  * @param spanProcessor
+  *   the [[SpanProcessor]] to call on span's end
+  *
+  * @param immutableState
+  *   the immutable state of the span
+  *
+  * @param mutableState
+  *   the mutable state of the span
+  *
+  * @tparam F
+  *   the higher-kinded type of a polymorphic effect
+  */
+private[trace] final class SdkSpanBackend[F[_]: Monad: Clock: Console] private (
+    spanProcessor: SpanProcessor[F],
+    immutableState: SdkSpanBackend.ImmutableState,
+    mutableState: Ref[F, SdkSpanBackend.MutableState]
+) extends Span.Backend[F]
+    with SpanRef[F] {
+
+  def meta: InstrumentMeta[F] =
+    InstrumentMeta.enabled
+
+  def context: SpanContext =
+    immutableState.context
+
+  def updateName(name: String): F[Unit] =
+    updateState("updateName")(_.copy(name = name)).void
+
+  def addAttributes(attributes: Attribute[_]*): F[Unit] =
+    updateState("addAttributes") { s =>
+      val next = Attributes.newBuilder
+        .addAll(s.attributes)
+        .addAll(attributes)
+        .result()
+
+      s.copy(attributes = next)
+    }.unlessA(attributes.isEmpty)
+
+  def addEvent(name: String, attributes: Attribute[_]*): F[Unit] =
+    for {
+      now <- Clock[F].realTime
+      _ <- addEvent(name, now, attributes: _*)
+    } yield ()
+
+  def addEvent(
+      name: String,
+      timestamp: FiniteDuration,
+      attributes: Attribute[_]*
+  ): F[Unit] =
+    addTimedEvent(
+      EventData(name, timestamp, Attributes.fromSpecific(attributes))
+    )
+
+  def recordException(
+      exception: Throwable,
+      attributes: Attribute[_]*
+  ): F[Unit] =
+    for {
+      now <- Clock[F].realTime
+      _ <- addTimedEvent(
+        EventData.fromException(
+          now,
+          exception,
+          Attributes.fromSpecific(attributes),
+          false
+        )
+      )
+    } yield ()
+
+  def setStatus(status: Status): F[Unit] =
+    updateState("setStatus") { s =>
+      s.copy(status = StatusData(status))
+    }.void
+
+  def setStatus(status: Status, description: String): F[Unit] =
+    updateState("setStatus") { s =>
+      s.copy(status = StatusData(status, description))
+    }.void
+
+  private[otel4s] def end: F[Unit] =
+    for {
+      now <- Clock[F].realTime
+      _ <- end(now)
+    } yield ()
+
+  private[otel4s] def end(timestamp: FiniteDuration): F[Unit] = {
+    for {
+      updated <- updateState("end")(s => s.copy(endTimestamp = Some(timestamp)))
+      _ <- toSpanData.flatMap(span => spanProcessor.onEnd(span)).whenA(updated)
+    } yield ()
+  }
+
+  private def addTimedEvent(event: EventData): F[Unit] =
+    updateState("addEvent")(s => s.copy(events = s.events :+ event)).void
```

Comment:
`events` is a `List`, isn't this pretty inefficient?

Reply:
Valid point. Let's use `Vector` under the hood.